src/webgpu/util/copy_to_texture.ts

```diff
@@ -32,34 +81,82 @@ export class CopyToTextureUtils extends GPUTest {
     return dstPixels;
   }
 
-  getExpectedPixels(
-    sourcePixels: Uint8ClampedArray,
-    width: number,
-    height: number,
-    format: RegularTextureFormat,
-    isFlipY: boolean,
+  getExpectedDstPixelsFromSrcPixels({
+    srcPixels,
+    srcOrigin,
+    srcSize,
+    dstOrigin,
+    dstSize,
+    subRectSize,
+    format,
+    flipSrcBeforeCopy,
+    srcDoFlipYDuringCopy,
+    conversion,
+  }: {
+    srcPixels: Uint8ClampedArray;
+    srcOrigin: GPUOrigin2D;
+    srcSize: GPUExtent3D;
+    dstOrigin: GPUOrigin3D;
+    dstSize: GPUExtent3D;
+    subRectSize: GPUExtent3D;
+    format: RegularTextureFormat;
+    flipSrcBeforeCopy: boolean;
+    srcDoFlipYDuringCopy: boolean;
     conversion: {
       srcPremultiplied: boolean;
       dstPremultiplied: boolean;
       srcColorSpace?: PredefinedColorSpace;
       dstColorSpace?: PredefinedColorSpace;
-    }
-  ): TexelView {
+    };
+  }): TexelView {
     const applyConversion = makeInPlaceColorConversion(conversion);
 
+    const reifySrcOrigin = reifyOrigin3D(srcOrigin);
+    const reifySrcSize = reifyExtent3D(srcSize);
+    const reifyDstOrigin = reifyOrigin3D(dstOrigin);
+    const reifyDstSize = reifyExtent3D(dstSize);
+    const reifySubRectSize = reifyExtent3D(subRectSize);
+
     const divide = 255.0;
     return TexelView.fromTexelsAsColors(
       format,
       coords => {
-        assert(coords.x < width && coords.y < height && coords.z === 0, 'out of bounds');
-        const y = isFlipY ? height - coords.y - 1 : coords.y;
-        const pixelPos = y * width + coords.x;
+        assert(
+          reifyDstOrigin.x + reifySubRectSize.width <= reifyDstSize.width &&
+            reifyDstOrigin.y + reifySubRectSize.height <= reifyDstSize.height,
+          'subrect is out of bounds'
+        );
+        assert(
+          coords.x >= reifyDstOrigin.x &&
+            coords.y >= reifyDstOrigin.y &&
+            coords.x < reifyDstOrigin.x + reifySubRectSize.width &&
+            coords.y < reifyDstOrigin.y + reifySubRectSize.height &&
+            coords.z === 0,
+          'out of bounds'
+        );
+        // Map dst coords to get candidate src pixel position in y.
+        let src_y = coords.y - reifyDstOrigin.y + reifySrcOrigin.y;
+
+        // If srcDoFlipYDuringCopy is true, a flipY op has been applied to src during copy.
+        // WebGPU spec requires origin option relative to the top-left corner of the source image, increasing downward consistently.
+        // https://www.w3.org/TR/webgpu/#dom-gpuimagecopyexternalimage-flipy
+        // Flip only happens in copy rect contents and src origin always top-left.
+        // Get candidate src pixel position in y by mirroring in copy sub rect.
+        if (srcDoFlipYDuringCopy)
+          src_y = reifySubRectSize.height - (src_y - reifySrcOrigin.y) - 1 + reifySrcOrigin.y;
```

Comment:
nit: Slight refactor to make this easier to understand (please review that it's correct):

```suggestion
        let yInSubRect = coords.y - reifyDstOrigin.y;

        // If srcDoFlipYDuringCopy is true, a flipY op has been applied to src during copy.
        // WebGPU spec requires origin option relative to the top-left corner of the source image, increasing downward consistently.
        // https://www.w3.org/TR/webgpu/#dom-gpuimagecopyexternalimage-flipy
        // Flip only happens in copy rect contents and src origin always top-left.
        // Get candidate src pixel position in y by mirroring in copy sub rect.
        if (srcDoFlipYDuringCopy)
          yInSubRect = reifySubRectSize.height - 1 - yInSubRect;

        let src_y = yInSubRect + reifySrcOrigin.y;
```